### ISSUE TITLE
GM via the built-in fluidsynth does not work on linux.

### ISSUE THREAD

**Issue by IlyaIndigo** [dosbox-staging/dosbox-staging#883]
**Describe the bug**
I specify in the config.
```
rate = 48000
midiconfig = 128:0
soundfont = /usr/share/sounds/sf2/FluidR3_GM.sf2
```

I start WarCraft and there is no music on the title screen.
If I run the fluidsynth daemon with the same soundfont, the music plays.
Maybe I am doing something wrong?

**To Reproduce**
1 Need linux with soundfont FluidR3
2 specify in the config.
```
rate = 48000
midiconfig = 128:0
soundfont = /usr/share/sounds/sf2/FluidR3_GM.sf2
```
3 Setup in WarCraft General Midi music card
4 Run WarCraft.

**Expected behaviour**
The music on the title screen should be perceived without an external demon.

**Environment (please complete the following information):**
- OS and version openSUSE Tumbleweed
- Relevant hardware info: Notebook Intel i3/Intel HD

**Version of dosbox-staging:**
0.76.0

**Configuration parameters:**
```
rate = 48000
midiconfig = 128:0
soundfont = /usr/share/sounds/sf2/FluidR3_GM.sf2
```

**Additional context**
https://build.opensuse.org/package/show/multimedia:apps/fluid-soundfont
https://build.opensuse.org/package/show/home:13ilya/dosbox-staging
[warcraft.log.txt](https://github.com/dosbox-staging/dosbox-staging/files/5973116/warcraft.log.txt)

**Comment by kcgen** [comment]
The log shows the configuration file is using ALSA as it's MIDI player (so you're using your kernel's sequencer for that): 

> MIDI: Opened device: alsa

@IlyaIndigo , can you try replacing your `[midi]` sections with the `[midi]` and `[fluidsynth]` sections described here:

https://dosbox-staging.github.io/v0-76-0/#2-3-add-built-in-midi-support-via-fluidsynth

**Comment by IlyaIndigo** [comment]
`mididevice = fluidsynth`
solves the problem, but with auto even after deleting midiconfig, dosbox-staging still tries to use ALSA and does not switch to fluidsynth.
`ALSA: Can't subscribe to MIDI port (65:0) nor (17:0)`

**Comment by IlyaIndigo** [comment]
Maybe with the auto option it is better to try to use mt32emu first, then if no ROM is found, then fluidsynth, and if no soundfont is found either, then use alsa with the specified port in midiconfig?
I was expecting this behavior.

**Comment by kcgen** [comment]
> Maybe with the auto option it is better to try to use mt32emu first, then if no ROM is found, then fluidsynth, and if no soundfont is found either, then use alsa with the specified port in midiconfig?
> I was expecting this behavior.

I do agree this feels backwards right now - by default, the build-in options should be used because 1) they're guarnateed to be there (if compiled in) and 2) we control their quality. Whereas external players aren't guaranteed to be present by the OS, not be optimally configured, and can be tricky to setup given the `midiconfig` in DOSBox also needs to be set properly. 

Currently, `auto` only iterates through external MIDI interfaces, and doesn't use either of the built-ins primarily for backward compatibility: DOSBox (vanilla) users expect to use their external MIDI synthesizer (Timidity++, BassMidi, FalconSoft, Roland-SC55, etc.. ) to play their MIDI audio.

Secondarily we know that FluidSynth can add a startup delay while is loads its SF2 file; so we didn't want to burden users with that delay on *every* startup, when we know lots of games don't even use MIDI.  We do have a PR that moves that work into the background while the splash screen is loading, so this is sufficient to eliminate that delay for modestly sized SF2.

Also, FluidSynth wasn't built-in to all of our 0.76 release binaries, but with 0.77 likely supporting both FluidSynth and MT-32 for all binaries, we might be able to reconsider switching this behaviour.

**Comment by IlyaIndigo** [comment]
Okay, I understand.
Thank you for your answer.

**Comment by IlyaIndigo** [comment]
@kcgen Do I understand correctly that the game can't tell dosbox which midi device is being used and if it's being used at all?

Then of course the default behavior of mt32emu (look for roms in the same directory as the executable) is very correct and logical.
If I want to use mt32, I create symlinks to roms and thereby communicate my choice to dosbox without changing the config.

Can we do the same for fluidsynth?
If the default "soundfont" option is empty, look for "FluidR3_GM.sf2" or "default.sf2" in the directory with the executable, and if it is not there or is incorrect, or the correct mt32 roms were previously connected, then do not activate built-in fluidsynth support.

True then you need to make a separate option for volume.
Strangely, I have to specify 25 to specify the same volume level as through the fluidsynth daemon.
Whether the default built-in fluidsynth is louder than the external one is good or bad, I don't know yet, I have no time to test the volume ratio in games.

**Comment by kcgen** [comment]
Correct; the game's MPU driver games sends sysex messages over port 330 to whatever device is running there (be it Roland MT32, CM32L, or a GM MIDI player). 

That's a very interesting suggestion regarding auto-selecting the device-type based on presence of either MT32 ROMs or an SF2 file in DOSBox's present PWD launch directorory. I agree this would be a very direct and convenient way to launch DOSBox, atleast on systems that support symlinks.

**Comment by kcgen** [comment]
> Whether the default built-in fluidsynth is louder

By default, the amplitude if the built in FluidSynth stream uses the SF2 native voice levels, whatever those might me, without any volume amplification or attenuation.  So it's largely a function of SF2.  Some can be quite loud (like Tibres of Heaven) while others more balanced (like SGM). 

Like the 0.76 release doc suggests, your `soundfont = filename` conf setting can include a post-fix volume scalar that's specific to that SF2 file. 

`soundfont = filename 50`
(Will set the gain to 50%).  For this setting, you want all your SF2s normalized against each other, so you can easily swap sf2 lines and the resulting levels will be consistent. 

You then have the `mixer fsynth %` setting too, which you can use in the game's launch batch file to set per-game desired music levels.

**Comment by dreamer** [comment]
OK, I can't address the whole discussion right now, but picking default MIDI backend is much more complicated than that…

- We cannot have MT-32 as default ANYWHERE, NEVER - because it depends on user installing non-free ROMs on their systems. MT-32 stays opt-in for as long as we won't have freely redistributable ROM replacements.
- Windows defaults to win32 midi backend (which does not require soundfont - it just works) and probably should stay this way.
- macOS default to CoreMIDI (?) and probably should stay this way as well
- FreeBSD defaults to OSS, but MIDI support of BSDs is generally extremely poor
- Linux ALSA should *probably* stay as default, but we *know* the client detection is broken (and was broken for years and years). I thought we had a bug report to track the fixing of this, but now I can't find it?
- Haiku and other OSes have their own MIDI backends which I am not interested in supporting, ever

GM via FluidSynth definitely *does* work on Linux - I use it all the time, with Fluid_R3 soundfont.
